File: backend/internal/models/notification_channel.go
package models

import (
	"encoding/json"
	"net/url"
	"regexp"
	"strings"
	"time"
)

// =====================================================================
// 外发通知通道 (设计/外发通知通道.md §3, 冻结数据模型)
// =====================================================================

// 通道类型 (预设模板标识, 设计 §2.4)。
// 预设只是前端表单的预填值: 三者共用同一套投递机制, 差异只在 URL 形态与请求体形状。
const (
	ChannelTypeWebhook = "webhook" // 通用 webhook (JSON POST)
	ChannelTypeWeCom   = "wecom"   // 企业微信群机器人 (key 在 URL 里)
	ChannelTypeOneBot  = "onebot"  // QQ 机器人 OneBot v11 (access_token 放 header)
)

// 通道最小级别 (MinLevel 取值; 与告警三级同名, 见 alert.go)。
// 语义: 只有级别 >= MinLevel 的通知才投递到该通道。
const (
	ChannelMinLevelInfo     = "info"
	ChannelMinLevelWarning  = "warning"
	ChannelMinLevelCritical = "critical"
)

// 投递状态 (NotificationDelivery.State 取值, 设计 §5 状态机)。
// pending 只在"尝试已开始、结论未落库"的窗口内存在 (崩溃窗口的审计证据);
// delivered 是成功终态; failed 表示该次尝试失败 (最后一次 failed 即整条通道投递的终态)。
const (
	DeliveryStatePending   = "pending"
	DeliveryStateDelivered = "delivered"
	DeliveryStateFailed    = "failed"
)

// NotificationChannel 外发通知通道 (通用 webhook)。
//
// 安全裁决 (设计 §2.3-b): Secret 是**只写不读**列 —— json:"-" 保证它
// 永不出现在任何 JSON 序列化结果里 (与 users.password_hash 同思路),
// 列表/详情接口只能回 SecretHint (末 4 位) 供用户辨认。
type NotificationChannel struct {
	ID   uint   `gorm:"primaryKey" json:"id"`
	Name string `gorm:"size:64;not null" json:"name"`
	Type string `gorm:"size:24;not null" json:"type"` // webhook | wecom | onebot（预设模板标识）
	// TargetURL 的出站地址**本身可能携带密钥**：企业微信群机器人的 URL 形如
	// https://qyapi.weixin.qq.com/cgi-bin/webhook/send?key=<secret>，
	// 密钥就在查询串里。因此它必须经 MarshalJSON 脱敏后才可出站到 API
	// （只保护 Secret 字段是不够的 —— 这正是 TestNotificationChannelSecretNeverSerialized
	// 抓到的缺陷：Secret 是 json:"-" 了，但 target_url 原样回显了 key）。
	TargetURL  string `gorm:"size:512;not null" json:"target_url"`            // 出站地址（序列化时脱敏）
	Secret     string `gorm:"size:512" json:"-"`                              // 只写不读（裁决 §2.3-b）
	SecretHint string `gorm:"size:8" json:"secret_hint"`                      // 末 4 位，供用户辨认
	Template   string `gorm:"type:text" json:"template"`                      // 请求体模板（JSON with {{.Title}} 等）
	MinLevel   string `gorm:"size:16;not null;default:info" json:"min_level"` // info|warning|critical
	// Enabled 同样**不带 gorm default** (设计 §7.6 的可复用结论): 带 default 的列
	// 无法表达"显式 false" —— GORM 把 bool 零值当"未设置"省略该列, DB 填回
	// default true, 于是"创建时就禁用的通道"会静默变成启用通道, 并且 GORM 还会把
	// DB 的默认值**回写进结构体**, 让调用方以为 false 生效了。
	// 默认值改由应用层给出 (见 NewNotificationChannel)。
	Enabled bool `json:"enabled"`
	// TimeoutSec / MaxRetries 是**指针**: nil = 未配置 (用默认值 10s / 2),
	// 0 = 用户显式要求的零值 (max_retries:0 就是"不重试")。
	//
	// 设计 §7.6 裁决: 带 gorm:"default:N" 的 int 列无法表达"显式零值" —— GORM 把
	// int 零值当"未设置"省略该列, DB 填回 default, 于是用户配的 max_retries=0 被
	// 静默改成 2 (投递耗时变 3 倍且无从察觉)。指针让 nil 与 0 可区分; 同时去掉
	// gorm default, 默认值只由应用层 DefaultTimeoutSec/DefaultMaxRetries 给出,
	// 不再存在"两处各说各话"的语义冲突。
	TimeoutSec *int `json:"timeout_sec"`
	MaxRetries *int `json:"max_retries"`
	// AllowPrivate 是设计 §7.1 冻结的 SSRF 例外开关: 家庭内网部署 OneBot
	// (http://192.168.x.x:5700/send_msg) 是合法场景。默认关闭; 开启后该通道
	// 可投递到 private/loopback/link-local/ULA 地址, 调用方与 UI 必须给出警示。
	AllowPrivate bool      `gorm:"default:false" json:"allow_private"`
	CreatedAt    time.Time `json:"created_at"`
	UpdatedAt    time.Time `json:"updated_at"`
}

func (NotificationChannel) TableName() string { return "notification_channels" }

// MarshalJSON 让通道在任何 JSON 出站路径上都不泄露凭据。
//
// 为什么必须在这里做，而不是在各 handler 里逐个处理：
//  1. 凭据有**两个**载体 —— `Secret` 字段（json:"-" 已挡住）与 `TargetURL` 的查询串
//     （企业微信 key=、OneBot access_token= 等）。只挡前者是漏的。
//  2. 通道会经多个出口被序列化：列表、详情、创建/更新响应、审计、WS 推送。
//     逐个 handler 处理必然会漏掉某一个 —— 放在类型上才是"一处定义"。
//
// 实现要点：先 Alias 出去拿默认序列化（避免与 json.Marshal 递归），
// 再把 target_url 换成脱敏版本。脱敏规则见 RedactTargetURL —— 它与
// notify.RedactURL 逐字节一致，占位符是 ***（与前端 stripToken 的占位一致）。
func (c NotificationChannel) MarshalJSON() ([]byte, error) {
	type alias NotificationChannel
	safe := alias(c)
	safe.TargetURL = RedactTargetURL(c.TargetURL)
	return json.Marshal(safe)
}

// RedactedPlaceholder 是 URL / 文案脱敏后的占位符，与 notify.RedactedPlaceholder
// （以及前端 stripToken 的占位）必须同值 —— 两处各有一份实现（原因见
// RedactTargetURL），唯一能防止再次漂移的办法就是"值只有一个定义、由测试断言相等"。
// notify 包的 TestRedactTargetURLMatchesNotify 里有一条断言专门守这个等式。
//
// 写成具名常量而不是字面量：占位符一旦被散布成魔法字符串，改一处漏一处就会
// 重新变成"两边输出不同"的契约漂移（P0-A 修的就是这个）。
const RedactedPlaceholder = "***"

// RedactTargetURL 是 models 包内的**最小脱敏实现**：同一输入下，输出与
// notify.RedactURL **逐字节相同**（由 notify 包的 TestRedactTargetURLMatchesNotify 守护）。
//
// 为什么不直接调用 notify.RedactURL：依赖方向是 notify → models（notify 要读
// NotificationChannel 才能投递），models → notify 会构成 import 环。
// 因此这里只能保留一份语义相同的副本。
//
// 由此还推出一条硬约束：**这个一致性测试不可能写在 models 包**——models 侧的
// 测试文件一旦 import notify，整个测试二进制在编译期就失败（import cycle）。
// 它只能待在 notify 包（internal/notify/redact_contract_test.go）。改本函数时
// 必须同步跑那条测试，否则契约会再次漂移成"注释声称有测试、实际没有"。
//
// 流水线与 notify.RedactURL 完全同构，三步缺一不可：
//  1. 空串早退；url.Parse 失败 ⇒ 走本包的文本级兜底 redactText（绝不原样返回）；
//  2. 剥掉 user:pass@ 用户凭据，并把敏感查询参数置为 RedactedPlaceholder；
//  3. query.Encode() 会把 '*' 百分号转义成 %2A，所以最后再套一遍 redactText，
//     把 key=%2A%2A%2A 收回成 key=***。
//
// 第 3 步**不能**简化成 strings.ReplaceAll(out, "%2A%2A%2A", "***")：那会把
// **非敏感参数里的用户数据**也一起改掉。反例 ?note=***&key=SECRET —— notify
// 输出 key=***&note=%2A%2A%2A（note 是用户数据，保持其百分号编码形态），
// 裸替换会把它篡改成 note=***。必须复用同一套正则，让"哪个值被改"只由参数名决定。
func RedactTargetURL(raw string) string {
	if strings.TrimSpace(raw) == "" {
		return ""
	}
	parsed, err := url.Parse(raw)
	if err != nil {
		// 解析失败绝不原样回显：非法 URL 里同样可能带 ?token=...（notify 侧同理）。
		return redactText(raw)
	}
	// 用户凭据 (user:pass@host) 一律剥掉。
	if parsed.User != nil {
		parsed.User = nil
	}
	if parsed.RawQuery != "" {
		query := parsed.Query()
		changed := false
		for name := range query {
			if sensitiveQueryKey(strings.ToLower(name)) {
				query.Set(name, RedactedPlaceholder)
				changed = true
			}
		}
		if changed {
			// url.Values.Encode 按 key 排序重编码（脱敏后的重编码，不要求与原文
			// 逐字节一致）；转义出来的 %2A 由下面 redactText 的兜底正则收回成 ***。
			parsed.RawQuery = query.Encode()
		}
	}
	return redactText(parsed.String())
}

// sensitiveQueryKeys 必须与 notify 包的 sensitiveQueryKeys **逐键一致**。
// 少一个键就是一条真实泄露面：notify 认 pwd / app_secret 而 models 不认时，
// ?pwd=SECRET 会在 API 响应里原样回显（P0-A 实测复现过的漂移）。
//
// 两个包的变量同名不冲突（不同包）；**故意保持同名**，好让"改一个必须改另一个"
// 在 diff 与审阅时一眼可见。守护测试会逐键对拍两张表。
var sensitiveQueryKeys = map[string]bool{
	"key": true, "token": true, "access_token": true, "accesstoken": true,
	"secret": true, "password": true, "passwd": true, "pwd": true,
	"sign": true, "signature": true, "sig": true, "auth": true, "authorization": true,
	"apikey": true, "api_key": true, "appkey": true, "app_secret": true,
}

func sensitiveQueryKey(lower string) bool { return sensitiveQueryKeys[lower] }

// redactURLRe / redactHeaderRe / redactBearerRe 是 notify 包同名正则的**等价副本**
// （import 环说明见 RedactTargetURL）。语义必须与 notify 侧逐字符一致：任何一侧
// 改了正则而另一侧没跟，TestRedactTargetURLMatchesNotify 就会变红。
var (
	// 兜底匹配任何形态 URL 里的敏感查询参数（含裸 query 串与 net/http 错误
	// 文案里被引号包住的 URL）。目标串: key=..., token=..., access_token=... 等，
	// 值到 &/#/空白/引号为止。
	redactURLRe = regexp.MustCompile("(?i)\\b(key|token|access_?token|secret|password|passwd|pwd|sign|signature|sig|api_?key|app_?key|app_?secret|auth|authorization)=([^&\\s\"'#]+)")
	// 兜底匹配 header 形态的凭据 (Authorization: xxx / X-Api-Key: xxx)。
	redactHeaderRe = regexp.MustCompile("(?i)((?:proxy-)?authorization|x-api-key|apikey)\\s*[:=]\\s*([^\\s,;\"']+)")
	// 兜底匹配 Bearer/Basic 方案里的令牌本体。
	redactBearerRe = regexp.MustCompile("(?i)\\b(bearer|basic)\\s+([A-Za-z0-9\\-._~+/=]{4,})")
)

// redactText 与 notify.RedactText 等价（说明见 RedactTargetURL）。
// 用途与 notify 侧一致：解析失败时兜底、以及把 Encode() 转义出的 %2A 收回成 ***。
func redactText(text string) string {
	if text == "" {
		return text
	}
	// 顺序有讲究: 先剥 Bearer/Basic 的令牌本体, 再处理 header 形态 ——
	// 反过来会把 "Bearer" 这个词当成值吃掉, 令牌本体就漏出去了。
	text = redactBearerRe.ReplaceAllString(text, "$1 "+RedactedPlaceholder)
	text = redactHeaderRe.ReplaceAllString(text, "$1: "+RedactedPlaceholder)
	return redactURLRe.ReplaceAllString(text, "$1="+RedactedPlaceholder)
}

// ApplySecret 写入密钥并同步辨认提示 (只写不读的唯一写入口)。
// secret 为空串表示清空密钥。
func (c *NotificationChannel) ApplySecret(secret string) {
	c.Secret = secret
	c.SecretHint = SecretHintOf(secret)
}

// SecretHintOf 取密钥末 4 位作为辨认提示 (设计 §3 SecretHint 注释)。
//
// 长度不足 4 位的密钥不回显原文, 一律以 '*' 占位 —— 否则"提示"就等于"密钥",
// 与 §7.3「密钥不落日志/不回显」冲突。长度为 0 时返回空串 (表示未设置)。
func SecretHintOf(secret string) string {
	if secret == "" {
		return ""
	}
	runes := []rune(secret)
	if len(runes) <= 4 {
		return strings.Repeat("*", len(runes))
	}
	return string(runes[len(runes)-4:])
}

// NewNotificationChannel 构造一个带显式默认值的通道。
// nil = 未配置 (由应用层归一化到设计默认值), 非 nil 的 0 是用户显式零值。
func NewNotificationChannel(name, channelType, targetURL string, timeoutSec, maxRetries *int) NotificationChannel {
	return NotificationChannel{
		Name:       name,
		Type:       channelType,
		TargetURL:  targetURL,
		MinLevel:   ChannelMinLevelInfo,
		Enabled:    true,
		TimeoutSec: timeoutSec,
		MaxRetries: maxRetries,
	}
}

// NotificationDelivery 投递审计 (每次尝试一行)。
//
// 为什么不用 notifications 上的标记列 (设计 §3): 一条通知可投多个通道、每个通道可重试多次,
// 是 1:N:M 关系; 压进一张表会丢失"哪个通道、第几次尝试、响应码多少"。
type NotificationDelivery struct {
	ID             uint64    `gorm:"primaryKey" json:"id"`
	NotificationID uint      `gorm:"index" json:"notification_id"`
	ChannelID      uint      `gorm:"index" json:"channel_id"`
	State          string    `gorm:"size:24;not null;index" json:"state"` // pending|delivered|failed
	AttemptNo      uint32    `gorm:"not null;default:1" json:"attempt_no"`
	StatusCode     int       `json:"status_code"`
	ErrorMessage   string    `gorm:"size:512" json:"error_message"`
	DurationMs     int64     `json:"duration_ms"`
	CreatedAt      time.Time `gorm:"not null;index" json:"created_at"`
}

func (NotificationDelivery) TableName() string { return "notification_deliveries" }

File: frontend-shared/src/api/notificationChannel.ts
import client, { type ApiEnvelope } from './client'

/**
 * 外发通知通道 API（后端 handler_notification_channel.go 的冻结契约：list/remove/create/update/test）。
 *
 * 为什么类型要逐字段手写而不是从后端生成：后端出站形状是 notificationChannelView
 * 这个**白名单**（handler_notification_channel.go:160），里头根本没有 secret 字段 ——
 * 前端类型里也不该有，这样"把明文密钥渲染出来"在类型层就不可能发生。
 */

/** 通道类型：与 models.ChannelType* 的三个预设标识一一对应。 */
export type NotificationChannelType = 'webhook' | 'wecom' | 'onebot'

/** 最低告警级别：后端 notify.MinLevelOf 会把未知值归一化成 info。 */
export type NotificationMinLevel = 'info' | 'warning' | 'error' | 'critical'

/**
 * 通道视图（= 后端 notificationChannelView 的逐字段投影）。
 *
 * 凭据只以 secret_hint（末 4 位）+ has_secret（布尔）出现：
 * 后者让前端能区分"没密钥"与"密钥太短导致 hint 为空"，不必去猜。
 */
export interface NotificationChannel {
  id: number
  name: string
  type: NotificationChannelType
  /** 已脱敏的目标地址（企业微信 key= 这类查询串凭据由后端 RedactTargetURL 处理） */
  target_url: string
  /** 密钥末 4 位；未设置时为空串 */
  secret_hint: string
  has_secret: boolean
  template: string
  min_level: NotificationMinLevel
  enabled: boolean
  /** null = 未配置（应用层默认）；0 = 用户显式零值（设计 §7.6，二者语义不同，不可折叠） */
  timeout_sec: number | null
  max_retries: number | null
  allow_private: boolean
  created_at: string
  updated_at: string
}

/** 列表查询参数（后端支持 type / enabled 过滤 + page / page_size 分页）。 */
export interface NotificationChannelListParams {
  type?: NotificationChannelType
  enabled?: boolean
  page?: number
  page_size?: number
}

/** 列表解包结果：后端 data 为 { items, total, page, page_size }。 */
export interface NotificationChannelListResult {
  items: NotificationChannel[]
  total: number
  page: number
  page_size: number
}

/** 删除响应：后端明确回 deliveries_retained=true（审计行**不**连带删除）。 */
export interface NotificationChannelRemoveResult {
  id: number
  deleted: boolean
  deliveries_retained: boolean
}

/**
 * 创建请求体（= 后端 createNotificationChannelRequest 的逐字段投影）。
 *
 * timeout_sec / max_retries 是 `number | null`：后端 DTO 是 *int，null 与 0 **语义不同**
 * （设计 §7.6）—— null = 未配置（应用层默认值生效），0 = 用户显式零值
 * （max_retries=0 就是"不重试"）。表单里"留空"必须发 null，绝不能发 0。
 *
 * secret 可省略：不传 = 该通道暂不配置密钥（例如仅用 URL 认证的企业微信）。
 */
export interface NotificationChannelCreatePayload {
  name: string
  type: NotificationChannelType
  target_url: string
  secret?: string
  template?: string
  min_level?: NotificationMinLevel
  enabled?: boolean
  timeout_sec?: number | null
  max_retries?: number | null
  allow_private?: boolean
}

/**
 * 更新请求体（= 后端 updateNotificationChannelRequest 的逐字段投影）。
 *
 * ★ secret 的三态是本文件最关键的契约（设计 §4 冻结 + §7.6 零值陷阱）：
 *   - **键不出现**（或值为 undefined）→ 后端 *string == nil → **不改密钥**，连 secret_hint 都不动；
 *   - `secret: ''`                        → 后端拿到非 nil 的空串 → **清空密钥**；
 *   - `secret: 'xxx'`                     → 覆盖为新密钥。
 *
 * 因此本类型里 secret 是 `?: string` 而**不是** `string`：类型层就要求调用方显式表达
 * "这次到底动不动密钥"。历史上把 secret_hint 当 secret 回传、或留空时回传空串，
 * 都会静默清掉用户已配置的密钥 —— 这正是本类型的注释要挡住的两类实现。
 *
 * 其余字段全部可选：后端 updateNotificationChannelRequest 是**全字段指针**（部分更新），
 * 未传 = 不动该列。这里若照抄创建体的必填字段，会凭空禁止合法的部分更新
 * （例如"只改个名字"），把后端能力误封在类型层。
 */
export type NotificationChannelUpdatePayload = Partial<Omit<NotificationChannelCreatePayload, 'secret'>> & {
  secret?: string
}

/** 测试投递响应：后端立即返回 pending，投递结果异步落到 notification_deliveries。 */
export interface NotificationChannelTestResult {
  channel_id: number
  notification_id: number
  /** 后端固定回 pending（models.DeliveryStatePending）——"已发出"不等于"已送达"。 */
  state: string
  /** 投递审计入口（投递审计页 NotificationDeliveries.vue 的查询入口）。 */
  deliveries_url: string
}

/**
 * 投递状态：与 models.DeliveryState* 一一对应（设计 §5 状态机 pending → delivered | failed）。
 *
 * pending 是**中间态**（"尝试已开始、结论未落库"）：后端先写一行 pending 再去出站，
 * 拿到结论后改写为 delivered/failed。因此页面上看到 pending 只表示"还在途中或进程
 * 在写入结论前退出了"，**不能**读成"已送达"，也不能读成"失败"。
 */
export type NotificationDeliveryState = 'pending' | 'delivered' | 'failed'

/**
 * 投递审计行（= 后端 models.NotificationDelivery 的逐字段投影，设计 §3）。
 *
 * 为什么类型要逐字段手写而不是从后端生成：后端出站形状就是这张**审计表本身**
 * （handler_notification_channel.go 的 listDeliveries 直接 Success(items)），
 * 表里根本没有任何密钥载体 —— 通道密钥在 notification_channels.secret 与 target_url
 * 的查询串里，审计行只有 error_message（落库前已过 RedactText）与 status_code。
 * 手写这九个字段，等于把"投递审计页拿不到密钥"这件事钉在类型层：页面想拼密钥也无处可取。
 */
export interface NotificationDelivery {
  id: number
  /** 被投递的通知 id（notifications 表） */
  notification_id: number
  /** 目标通道 id；通道被删除后审计行仍在（DELETE 不级联），故这里可能指向已不存在的通道 */
  channel_id: number
  state: NotificationDeliveryState
  /** 第几次尝试，从 1 开始；重试**新开一行**、attempt_no 递增（不是原地改） */
  attempt_no: number
  /** 出站 HTTP 状态码；0 = 没走到拿到响应那一步（连不上/DNS/SSRF 拒绝/超时） */
  status_code: number
  /** 失败原因（后端已脱敏并截断）；成功时为空串 */
  error_message: string
  /** 本次尝试耗时（毫秒） */
  duration_ms: number
  /** 审计行创建时刻（= 该次尝试开始时刻） */
  created_at: string
}

/** 投递审计查询参数（后端支持 channel_id / state 过滤 + page / page_size 真分页）。 */
export interface NotificationDeliveryListParams {
  /** 只看某条通道的投递；不传 = 全部通道 */
  channel_id?: number
  /** 只看某种状态；不传 = 全部状态 */
  state?: NotificationDeliveryState
  page?: number
  page_size?: number
}

/** 投递审计解包结果：后端 data 为 { items, total, page, page_size }（与通道列表同形）。 */
export interface NotificationDeliveryListResult {
  items: NotificationDelivery[]
  total: number
  page: number
  page_size: number
}

/** 拦截器返回后端统一 envelope；只从 envelope.data 取值（同 api/dataSource.ts 范式）。 */
async function unwrap<T>(p: Promise<ApiEnvelope<T>>): Promise<T> {
  return (await p).data
}

export const notificationChannelApi = {
  async list(params?: NotificationChannelListParams): Promise<NotificationChannelListResult> {
    return unwrap<NotificationChannelListResult>(
      client.get<unknown, ApiEnvelope<NotificationChannelListResult>>('/api/v1/notification-channels', { params }),
    )
  },
  async remove(id: number): Promise<NotificationChannelRemoveResult> {
    return unwrap<NotificationChannelRemoveResult>(
      client.delete<unknown, ApiEnvelope<NotificationChannelRemoveResult>>(`/api/v1/notification-channels/${id}`),
    )
  },
  async create(payload: NotificationChannelCreatePayload): Promise<NotificationChannel> {
    return unwrap<NotificationChannel>(
      client.post<unknown, ApiEnvelope<NotificationChannel>>('/api/v1/notification-channels', payload),
    )
  },
  /**
   * 更新通道（部分更新：只有出现在 payload 里的键才会被后端写库）。
   *
   * 调用方必须自己保证"留空的密钥"= **键不出现**：这里不做任何 JSON.stringify 层面的
   * 兜底（例如删掉 undefined 键）—— 那是把契约藏在看不见的地方；改为让类型
   * （NotificationChannelUpdatePayload.secret?: string）与调用点显式表达。
   */
  async update(id: number, payload: NotificationChannelUpdatePayload): Promise<NotificationChannel> {
    return unwrap<NotificationChannel>(
      client.put<unknown, ApiEnvelope<NotificationChannel>>(`/api/v1/notification-channels/${id}`, payload),
    )
  },
  /**
   * 发送测试消息（POST /:id/test）。
   *
   * 后端是**异步投递**：HTTP 200 只代表"测试消息已发出"，出站结果稍后落到投递审计
   * （响应 state 恒为 pending）。因此 UI 文案必须是"已发出"而不是"投递成功" ——
   * 这是本仓"假绿"纪律的直接应用：不能把"请求成功"说成"送达成功"。
   */
  async test(id: number): Promise<NotificationChannelTestResult> {
    return unwrap<NotificationChannelTestResult>(
      client.post<unknown, ApiEnvelope<NotificationChannelTestResult>>(`/api/v1/notification-channels/${id}/test`),
    )
  },
  /**
   * 投递审计列表（GET /notification-deliveries）。
   *
   * 服务端**真分页**：page / page_size 原样发给后端，返回的 items 就是**当前页切片**，
   * total 是全量总数（后端在 Offset/Limit 之前先 Count）。调用方不得对 items 再切片 ——
   * 那样页码与数据会双双脱节（分页器说第 2 页、内容却是第 1 页的前 N 条）。
   *
   * 参数不传即"不过滤"：channel_id / state 都省略时返回全部通道的全部审计行。
   * state 只能是 pending|delivered|failed —— 后端对非法值直接 400，前端类型与
   * 下拉项都不产生别的取值（不在这里做静默兜底，否则非法值会变成"看起来过滤了"）。
   */
  async listDeliveries(params?: NotificationDeliveryListParams): Promise<NotificationDeliveryListResult> {
    return unwrap<NotificationDeliveryListResult>(
      client.get<unknown, ApiEnvelope<NotificationDeliveryListResult>>('/api/v1/notification-deliveries', { params }),
    )
  },
}

export default notificationChannelApi
